staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v1/etag.go

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"crypto/sha512"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/endpoints/handlers/negotiation"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+)
+
+// This file exposes helper functions used for calculating the E-Tag header
+// used in discovery endpoint responses
+
+// Unsure if there is an existing interface for this
+type Marshalable interface {
+	Marshal() ([]byte, error)
+}
+
+// Attaches Cache-Busting functionality to an endpoint
+//   - Immutability Response Header
+//   - Expires Never
+//   - Redirects when incorrect hash is provided
+//   - Sets ETag haeader to provided hash
+//   - Replies with 304 Not Modified, if If-None-Match header matches hash
+//
+// hash should be the value of CalculateETag on object. If hash is empty, then
+//
+//	the object is simply serialized without E-Tag functionality
+func ServeHTTPWithETag(
+	object runtime.Object,
+	hash string,
+	serializer runtime.NegotiatedSerializer,
+	w http.ResponseWriter,
+	req *http.Request,
+) {
+	reqURL := req.URL
+	if reqURL == nil {
+		// Can not find contract guaranteeing any non-nility of req.URL
+		w.WriteHeader(500)
+		return
+	}
+
+	if len(hash) > 0 {
+		if providedHash := reqURL.Query().Get("hash"); len(providedHash) > 0 {
+			if hash == providedHash {
+				// The Vary header is required because the Accept header can
+				// change the contents returned. This prevents clients from
+				// caching protobuf as JSON and vice versa.
+				w.Header().Set("Vary", "Accept")
+
+				// Only set these headers when a hash is given.
+				w.Header().Set("Cache-Control", "public, immutable")
+
+				// Set the Expires directive to the maximum value of one year from the request,
```

Comment:
how about one day instead?  It's still a long time, but if have a bug/problem it can still correct on a practical time scale.